src/pystack/_pystack/pyframe.cpp

```diff
@@ -44,14 +44,33 @@ FrameObject::FrameObject(
     d_prev = nullptr;
     d_next = nullptr;
 
+    d_is_shim = false;
```

Comment:
The constructor is getting a bit complicated. What do you think if we start making small private functions to absorb the version-specific stuff something like:

```
uintptr_t FrameObject::getLastInstruction(const std::shared_ptr<const AbstractProcessManager>& manager,
                                          PyFrameObject& frame) {
    if (manager->majorVersion() > 3 || (manager->majorVersion() == 3 && manager->minorVersion() >= 11)) {
        return manager->getField(frame, &py_frame_v::o_prev_instr);
    } else {
        return manager->getField(frame, &py_frame_v::o_lasti);
    }
}

void FrameObject::checkForShim(const std::shared_ptr<const AbstractProcessManager>& manager,
                               PyFrameObject& frame) {
    d_is_shim = false;
    if (manager->majorVersion() > 3 || (manager->majorVersion() == 3 && manager->minorVersion() >= 12)) {
        constexpr int FRAME_OWNED_BY_CSTACK = 3;
        d_is_shim = manager->getField(frame, &py_frame_v::o_owner) == FRAME_OWNED_BY_CSTACK;
    }
}

void FrameObject::checkForEntry(const std::shared_ptr<const AbstractProcessManager>& manager) {
    if (manager->majorVersion() > 3 || (manager->majorVersion() == 3 && manager->minorVersion() >= 11)) {
        d_is_entry = false;  // Entry frames are detected below
    } else {
        d_is_entry = true;   // All frames are entry frames
    }
}
```

not sure if is worth it or is the time to do it though

Reply:
OK. I've added a commit that restructures this constructor to be implemented in terms of static helper functions that handle each of the version-specific things (getting the code object, checking if it's a shim frame, getting the previous frame, getting the `is_entry` flag). I think it's quite a lot better now. Let me know if you like it.